tests/request_helpers_test.py

```diff
@@ -0,0 +1,408 @@
+                    '-'.join(('a', '')),
+                    '-'.join(('', 'b')),
+                    '-'.join(('a', '', 'whee')),
+                    '-'.join(('a', span_id, 'whee')),
+                    '-'.join(('a', span_id, 'whee', parent_span_id)),
+                    '-'.join(('a', span_id, 'whee', '')),
+                    ]:
+        assert None is request_helpers.extract_zipkin_attrs_from_headers({
+            'b3': invalid,
+        }, sample_rate=88.2), invalid
+        assert [] == mock_random.mock_calls
+
+        # Convert the (bad) input b3 header into multiple-header format input
+        bits = invalid.split('-')
+        if len(bits) == 1:
+            bad_headers = {'X-B3-TraceId': bits[0]}
+        elif len(bits) == 2:
+            bad_headers = {'X-B3-TraceId': bits[0],
+                           'X-B3-SpanId': bits[1]}
+        elif len(bits) == 3:
+            bad_headers = {'X-B3-TraceId': bits[0],
+                           'X-B3-SpanId': bits[1],
+                           'X-B3-Sampled': bits[2]}
+        elif len(bits) == 4:
+            bad_headers = {'X-B3-TraceId': bits[0],
+                           'X-B3-SpanId': bits[1],
+                           'X-B3-Sampled': bits[2],
+                           'X-B3-ParentSpanId': bits[3]}
+
+        print('bad_headers: %r' % bad_headers)
+        assert None is request_helpers.extract_zipkin_attrs_from_headers(
+            bad_headers, sample_rate=88.2,
+        )
+        assert [] == mock_random.mock_calls
+
+    # I'm pretty sure a provided X-B3-Sampled with empty-string "value" should
+    # be considered invalid (bit of an edge case).
+    assert None is request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': 'a',
+        'X-B3-SpanId': span_id,
+        'X-B3-Sampled': '',
+        'X-B3-ParentSpanId': parent_span_id,
+    }, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+
+@pytest.mark.parametrize("no_trace_str,yes_trace_str",
+                         [('0', '1'), ('false', 'true')])
+@mock.patch('py_zipkin.util.random.random')
+def test_extract_zipkin_attrs_from_headers_multi(mock_random, no_trace_str,
+                                                 yes_trace_str):
+    trace_id = 'not_actually_validated'
+    span_id = '7e5991634df0c66e'
+    parent_span_id = '987d98239475e0a8'
+
+    # No B3-headers; return None to get a new local-root span
+    assert None is request_helpers.extract_zipkin_attrs_from_headers(
+        {}, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+    # Just 2 bits is a sample-rate defer; die-roll is "no":
+    mock_random.return_value = 0.883
+    assert ZipkinAttrs(
+        trace_id, span_id, None, '0', False,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+    }, sample_rate=88.2)
+    assert [
+        mock.call()
+    ] == mock_random.mock_calls
+    mock_random.reset_mock()
```

Comment:
why do you care? The response is correct and I don't think these extra asserts add much value and make the tests even longer

Reply:
Fair point; removing.